### PR TITLE
SQL storage provider by @stascode 

### PR DESCRIPTION
**Pull request by sergeybykov**
Initial version of the SQL storage provider that was created by @stascode as a starting point sample.

Refactoring and redesign ideas are very welcome.

### CODE REVIEW

**Comment by jthelin** [comment]
Does this need to be in the core runtime codebase?

We already have some contributed storage providers in Orleans-Contrib -- is this in a different equivalence class then those? 
How much does this stand alone and build able as a layer on top of core runtime packages?

**Comment by veikkoeeva** [comment]
@jthelin A good question. Maybe one way of thinking is that should all storage providers be in a contribution project?

Some observations:
* Is [Microsoft.Azure.SqlDatabase.ElasticScale.Client](https://github.com/dotnet/orleans/pull/959/files#diff-0ff0b22508cbf71bb1d611e06c2fade4R41) (also [Elastic Database features overview](https://azure.microsoft.com/en-us/documentation/articles/sql-database-elastic-scale-introduction/)) needed or is the intended functionality the same as [ShardedStorageProvider](https://github.com/dotnet/orleans/blob/master/src/OrleansProviders/Storage/ShardedStorageProvider.cs)?
* Is it Orleans that handles transient failures ([looking at the PR](https://github.com/dotnet/orleans/pull/959/files#diff-c5237c17b0bbe04354eab5da181f85a3R232)) or the storage provider?
* It looks like the database objects in this PR are [purpose-built](https://github.com/dotnet/orleans/pull/959/files#diff-18e3749cc5dfe89762c16e76ae47dd99R20) for certain grain state classes...
 * ... Which gives a chance to further discuss the design. Maybe this could be generalized so that either the grain type and ID are separate columns or combined in one according to Orleans formatting convention and the data column is of type `VARBINARY(MAX)`. This way any type of state class can be serialized without further action besides generating the database. This data column could be also of type `NVARCHAR(MAX)` (works also for [JSON](http://blogs.msdn.com/b/jocapc/archive/2015/05/16/json-support-in-sql-server-2016.aspx) in SQL Server 2016) or `XML` or all three could be present, but only one to be used with a given ID and type combination of an object.
 * Which way data is serialized depends of the choice of a serializer. I believe Orleans Storage Providers do not support the choice of a serializer, or usign a custom serialization function, at the moment, so, it'd be hardcoded at first. My take would be to use `VARBINARY(MAX)` and a binary serializer for the sake of efficiency (with [streaming support](http://blogs.msdn.com/b/adonet/archive/2012/07/15/using-sqldatareader-s-new-async-methods-in-net-4-5-beta-part-2-examples.aspx) added to relation provider code).
* I think there is value in being able to lay down data to purpose-built table structure and perhaps even doing processing on the storage bypassing application level. For the same mechanism would work as in other parts where Orleans loads the queries by key from the database. In combination with a method to choose a custom serializer function, one could re-write the query in the database so it would check from the incoming data the grain type (and other relevant information) and simply in switch-case choose to write the data in some other way than to the aforementioned data column, reading would be patched to match writing.
* Should storage data schema have an explicit version number, i.e. a separate column, or should state data be versioned in the data itself?
* Currently the ADO.NET relational code needed to support queries is in SqlUtils, but I suppose this is not a problem.

I think some of these questions are more general across all storage providers, like choosing a serialization format and versioning. Some applicable only to relational storage provider, such as how to load queries while there are differences between vendor products and versions and one might want to make use of some vendor and deployment specific features (the reason database licenses sell) and *maybe* sharding options.

Tangential observation regarding documentation [Custom Storage Providers](http://dotnet.github.io/orleans/Step-by-step-Tutorials/Custom-Storage-Providers) documentation that it mentions "OrleansSDK", which does not exist in newer versions.

**Comment by gabikliot** [comment]
The two reasons to include mature providers in the core and not in the contrib, for my opinion, are:
1) improved visibility - everyone using Orleans will discover them immediately.
2) improved quality - as more people will be seeing and using them, more bugs will be discovered and more advanced features added. For example, putting SQL storage provider next to SQL system store means more eyes on both (evidence is veiko's excellent comments above).

If course, we shouldn't include anything, but this specific one looks like meets the bar. The next ones could be blob provider and doc db provider.

And of course, all *utils assemblies shouldn't have internal access to core, so they should indeed be a layer on top of the core.

**Comment by stascode** [comment]
Keep in mind that the main goal was to bound number of SQL connections from Olreans Silo to a SQL DB. So SQL TVP (Table View Parameters) were used to do batched reads and writes from/to a SQL DB.
Feel free to modify the code according to your needs.

**Comment by sergeybykov** [comment]
@veikkoeeva I updated the [Custom Storage Providers](http://dotnet.github.io/orleans/Step-by-step-Tutorials/Custom-Storage-Providers) documentation page.

**Comment by veikkoeeva** [comment]
@stascode Ackowledged. Can you share more details on the reasons on using sharding with this customer? I can foresee a connection pool depletion being a problem if hundreds or thousands of stateful grains come alive at once and they are all individually loading data from the db. I see the importance of batched inserts for maintaining performance, but do you mean data was first loaded to TVPs over multiple inserts and then from there pumped to tables?

Batched inserts was a problem I was thinking when writing statistics insertion code. Not all databases have TVPs, so I wrote a piece of code to create a parameterized (can be non-parameterized too) `INSERT INTO UNION ALL` (in case of Oracle, `FROM DUAL` is added appropriately) dynamically according to how many parameters there are in an input collection. The [hand-tuned version](https://github.com/dotnet/orleans/blob/538e579e926ca7a12c310be2e3fa978e95a4ee6c/src/OrleansSQLUtils/Storage/OrleansRelationalExtensions.cs#L188) and one [using reflection](https://github.com/dotnet/orleans/blob/538e579e926ca7a12c310be2e3fa978e95a4ee6c/src/OrleansSQLUtils/Storage/OrleansRelationalExtensions.cs#L188).

**Comment by veikkoeeva** [comment]
@sergeybykov :+1:

**Comment by sergeybykov** [comment]
@jthelin SQL storage provider is meant to fill the gap we have in our SQL support story. Several people repeatedly asked for it.

In general, I'm of the opinion that we should bring high quality providers into the Orleans repo for, echoing @gabikliot's reasoning, better quality and discoverability. The code needs to meet a certain quality bar to be brought in, so that the experience of somebody using it as part of the Orleans codebase meets their expectations.

Provider code sitting in separate repos decays over time as Orleans evolves. Keeping it inside the repo will ensure that the code at least gets updated with any future API changes.

**Comment by sergeybykov** [comment]
@jthelin I removed the versions from project assembly references.

**Comment by jthelin** [comment]
Yes, i think it is useful to distinguish between "same repo" and "same code tree".

I also agree with your comments about code-rot -- we have already seen that happen with the Samples code because we don't regularly build those. :(

In addition, it helps to keep us honest if the user level "plugins" [without access to runtime internals] are built separately from the core Orleans runtime code, just like external developers would have to do. That helps to ensure we have the NuGet packaging right too.

So perhaps we should create a new top level `Providers` directory inside the Orleans repo, but separate from runtime src tree? 
That could include this new SQL Azure provider initially, and then eventually move the existing SQL provider and ZK projects under there too? 
Longer term we could move the AzureBlob provider, etc in there too, if the authors are amenable.

In that model of the repo structure, then each provider (singularly or grouped) would have their own ReadMe (usage instructions / pointers), test cases and build their own NuGet packages which would get included in the larger Orleans NuGet collection.

This brings up another question for this change - which NuGet package is this included / distributed in? 
Should it be added to the OrleansSqlUtils ..nuspec or its own OrleansSqlAzureProvider .nuspec?

I agree that having a good base collection of recommended providers and plugins is a good thing :)

File: src/Orleans.sln [hunk 1]
```diff
@@ -228,5 +228,8 @@ Global
 	GlobalSection(NestedProjects) = preSolution
 		{0BB483EF-6401-424C-9C4F-5750B300C7FD} = {DD5D35C0-A5FE-41CC-A10D-CAAA37996479}
 	EndGlobalSection
+	GlobalSection(ExtensibilityGlobals) = postSolution
+		EnterpriseLibraryConfigurationToolBinariesPathV6 = packages\EnterpriseLibrary.TransientFaultHandling.6.0.1304.0\lib\portable-net45+win+wp8
```
Comment (jthelin) [inline src/Orleans.sln]: Can someone explain what this bit is for, and how it is different than having package `EnterpriseLibrary.TransientFaultHandling` listed in `src/OrleansSQLUtils/packages.config` ?

Comment (jthelin) [inline src/Orleans.sln]: Mostly just curious to know what NuGet Manager is trying to do here, and why it decided that `EnterpriseLibrary.TransientFaultHandling` package requires special treatment.

**Comment by sergeybykov** [comment]
>So perhaps we should create a new top level  Providers  directory inside the Orleans repo, but separate from runtime src tree? Size and build time aren't even remotely an issue.

What's the benefit of doing that? I think keeping quality providers within Orleans.sln is better because that will guarantee they get built every time. The internal visibility is done by assembly name. It's independent from whether or not the friend assembly is in in the same source tree. So we'd have to ensure that regardless.

>So perhaps we should create a new top level  Providers  directory inside the Orleans repo, but separate from runtime src tree? 
 That could include this new SQL Azure provider initially, and then eventually move the existing SQL provider and ZK projects under there too? 
 Longer term we could move the AzureBlob provider, etc in there too, if the authors are amenable.

>In that model of the repo structure, then each provider (singularly or grouped) would have their own ReadMe (usage instructions / pointers), test cases and build their own NuGet packages which would get included in the larger Orleans NuGet collection.

I think these providers should be under src\Providers instead, included in Orleans.sln. They can still have individual readmes, NuGet packages, and whatnot.

>This brings up another question for this change - which NuGet package is this included / distributed in? 
 Should it be added to the OrleansSqlUtils ..nuspec or its own OrleansSqlAzureProvider .nuspec?

I think OrleansSqlUtils is a reasonable place for it. I don't see much value in splitting OrleansSqlUtils  into individual NuGets for storage provider, membership table provider, statistics table provider, etc.

**Comment by jthelin** [comment]
Having extra providers in Orleans.sln means it takes [much] longer to open and/or build that mega-solution, but does mean that any API renames (but hopefully this should be very infrequent now) and/or find usage is easier.
I agree that having the "extras" under `src\Providers\Storage` etc would make it clearer about their role and relationship to the other code. 
Who knows, maybe we could add a `src\Providers\Replication` tree at some point in the near future ;)

OTOH, we can guarantee that any standalone Providers tree would get built each time by adding call to `Providers\Build.cmd` in the top level `Build.cmd` script just after the call to `src\Build.cmd`. Similar thing for Test.cmd. 

We could (and really should) also add call to `Samples\Build.cmd` in the top level build script too, although dependencies on Node.js and Visual Basic in Samples means that is not completely working (yet) That would prevent the periodic code-rot in Samples. :(

I can tell you from painful recent experience that once you get up to 48 projects and ~500 files in one .sln, you will definitely [a] feel like you want to split into smaller build units, and [b] end up taking way more coffee breaks per day during rebuilds than is probably healthy! 
My rule of thumb is that keeping under 20 projects and 500-600 source files per .sln is "mostly bearable". 
Orleans is currently at 17 projects and 603 files, which didn't just happen by magic if you recall. ;)

**Comment by jthelin** [comment]
@sergeybykov - For consistency, can the following classes / interfaces be made internal (like most others in this change set) rather than public?
- GrainIdentity
- GrainStateMap
- IGrainStateMapFactory
- ReadEntry
- WriteEntry

I don't think the intent here is for these to be generally re-usable outside this provider, so let's try not to surprise or confuse people with new public APIs ;)

Then need to rebase this PR against current HEAD and i think we should be ok to merge to master?

**Comment by sergeybykov** [comment]
@jthelin You are right I was able to make them internal after a little bit of massaging. Done, rebased, and pushed.

**Comment by veikkoeeva** [comment]
A note: The storage layout is [this](https://github.com/dotnet/orleans/pull/959/files#diff-18e3749cc5dfe89762c16e76ae47dd99R51), saving [these](https://github.com/dotnet/orleans/pull/959/files#diff-ab4731b3ac530a7b666aa78c1557b035R8) which are not that general purpose. I think soon enough some wants to refactor that part, maybe even me. :)

**Pull request merged**